### DISCUSSION TITLE
Using ansible-pull

### DISCUSSION THREAD

**Discussion by erkrali** [bluebanquise/bluebanquise#899] [Ideas]
I see that there seems to be a work in progress to somewhat detect when a node that was just PXE installed just booted for the first time, and then automatically run `ansible` on it.

What about using [`ansible-pull`](https://docs.ansible.com/ansible/latest/cli/ansible-pull.html) instead? It seems to be designed for scaling out in this kind of scenarios.

I have managed to make it work from a preseed file for debian (not exactly trivial, but not too hard either. In my case, I have used [`git-http-backend`](https://git-scm.com/book/en/v2/Git-on-the-Server-Smart-HTTP) and the following preseed `late_command`:

```
  in-target sh -c 'if [ -n "$playbook" ]; then mkdir /dev/shm; mount none /dev/shm -t tmpfs -o nosuid,nodev; ANSIBLE_CONFIG=bluebanquise/ansible.cfg ANSIBLE_COLLECTIONS_PATH=collections ansible-pull --full --accept-host-key --purge -i inventory -U http://$ipxe_next_server/git/cluster-infrastructure playbooks/$playbook; umount /dev/shm; rmdir /dev/shm; fi'; \
```

I have to say the hardest part was to make your collections available without installing them through `ansible-galaxy`, I created a separate repository that has yours as a submodule and made some symlinks...

**Comment by oxedions** [comment]
Dear @erkrali 

Many thanks for this proposal !

I investigate the ansible-pull a long time ago to be honest.
The main issues at that time were:

1. How users can add their own roles or custom galaxies, but it seems you solved it.
2. Security. Ansible-pull assumes clients can access the whole inventory. This means everyone can read inventory content, and so if a user can ssh on the cluster, he/she could be able to git pull the inventory and grab secrets.
3. How to manage ansible vaults (some BlueBanquise users are using vaults).

However, your idea is interesting. Most clients nodes do not need to have access to the whole inventory. There could be some way to only expose a part of the inventory. I need to think about it. We could provide an optional ansible-pull step in autoinstallation files :slightly_smiling_face:

**Comment by erkrali** [comment]
Hi!

About 1., it's all about the directory structure... I can share my solution, but it's quite specific to my use case, so I'm not sure how useful it would be generically. In short, I created a rep where I keep my playbook, roles, etc., and then added bluebanquise as a subrepo. Then you only need to create `collections/ansible_collections/bluebanquise` as a symbolic link to `../../bluebanquise/collections` (and use the environment variable as appropriate).

For the security, what I did is to serve a copy of the repo from the manager node itself using git-http-backend. This copy can be cloned anonymously, unlike the "normal" one, but to compensate for that the nodes block access to the deployment server for non-root users.

I created a specific role for that that uses `nftables` directly, because as far as I know [it's not possible to do it with firewalld](https://github.com/firewalld/firewalld/issues/725).

Now I have no experience with vaulst, so I don't know what solution, if any, there might be.

But in any case, if there's the option to add custom content to the installation files (preseed, kickstart, etc.) I think this is something that can also be managed individually for each use case, I just wanted to start the discussion in case there was general interest in the feature.

**Reply by oxedions** [reply]
Dear @erkrali 

Thank you for this reply.

Hum, your idea of using nftables to filter access is a smart move. This could solve the issue.

There are many users interested by your idea. I had many requests lately on how to go beyond the current limits of cluster size. Basically, whatever I do to optimize my templates and tasks, execution speed is deeply related to CPU power of the management node that do the push. Biggest known cluster to me running BlueBanquise is 1400 HPC servers, and deployment of all nodes at once takes a long long time.

I found ways to greatly improve performances. Reduce calls to hostvars in the code, use buffers, do not use ansible-playbook -f but instead fork it by chunks of threads (so instead of doing an ansible-playbkk -f 32 for example, do 4 ansible-playbok -f 8 --limit foo), etc.
I also discovered that network speed is absolutely not related to Ansible execution (I compared an 100mb Eth to an Infiniband HDR to deploy a 1200 servers clusters, no changes (as long as you do not install packages, but this is unrelated to Ansible execution itself)).

But at some point, even a huge dual AMD EPIC server cannot deploy an ultra marge cluster without admin to have go take a coffee or even a lunch.

Providing the ansible-pull mechanism could be a huge improvement.

1. Large provisioning usually occur only during maintenance frames or initial deployment, and so cluster is considered safe at these times.
2. Admin can disable ansible-pull on demand, by closing ports or stopping git http server, and rely on standard playbooks to later provision isolated nodes (hardware replacement, etc.)

For BlueBanquise to propose this ansible-pull feature, there would be the following needs:

1. Be able to share the playbooks (you did it)
2. Be able to share the inventory (you did it)
3. Be able to restrict access to inventory to some specific nodes (you did it)
4. Add ansible-pull capability both in autoinstallation file and in direct usage (documentation mainly)
5. Tag reboot tasks to prevent a reboot in autoinstall
6. Find a simple way to distribute ansible on client nodes (cannot rely on pip for air gapped clusters, so I would have to provide an rpm that contains a Python venv with ansible, not a big deal)

I propose to start working on that once current code is stabilized (there are still too many issues to fix in current code due to introduction of Debian/Ubuntu support, and split of ep to os and hw, but this should solve soon) :blush:

**Comment by erkrali** [comment]
Hi!

I have been thinking a bit more about this issue, and here are some further ideas:

About distributing ansible (need 6. in your list), at least in debian, in my simple case it was enough to add it to the list of packages being installed. Admittedly, I wasn't using (yet) many bluebanquise roles for the "worker node", so I might still face compatibility issues later.

Now, I have rather been thinking about the security angle. The firewall solution is enough for my needs, but for environments where the security needs to be tighter I have thought a couple more options:

1. Boot the nodes from virtual media. In a cluster it's highly likely that you'll have either redfish or some other mass management tool, so you can create some .iso file that just boots to iPXE, only adding some "secret" (it can be either the HTTP credentials, or some password to a vault, or even a private key to decrypt some content).
2. Keep the "node booted" detection system, but instead of running ansible from the deployment server, just copy over a file with the necessary secrets (and permissions), and then call `ansible-pull` from the node. This should be feasible if you're already able to use ansible, and the main difference would be that sending the small file should be almost no cost, and ansible would be run in parallel from each node, so no load for the deployment server.
3. The "ideal" and cool way for the really security-oriented would be to pre-load the secrets in the TPM module (or eventualy UEFI variables, or some other similar place). Unfortunately this doesn't seem to be possible from redfish [1]... It seems somebody already had this exact same idea and didn't work.

So, as I said, I'm happy with the firewall, but just a bit more brainstorming in case it ends up being useful.

[1]: https://redfishforum.com/thread/513/leaving-breadcrumbs-host-redfish-api

**Reply by oxedions** [reply]
Dear @erkrali 

My apologies for the delay of my answer. Father duties are using all my free time :blush: 

Thank you very much for these idea. Once all the urgent issues are closed, I will work on this. This is very interesting and I would really like to be able to provide both ways to provision nodes. If no major bugs are opened, I should be able to work on that soon :slightly_smiling_face:
